src/NavigationStack.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2024 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+import { JsonPaletteType } from "./index.d";
+
+export class NavigationStack {
+
+  // The actual stack keeping track of where the user was.  The top-most palette
+  // is the most recent one the user was at before navigating to a new
+  // layer/palette.
+  navigateBackStack: Array<JsonPaletteType>;
+
+  // The current palette in the palette display area
+  currPalette: JsonPaletteType;
+
+  /**
+   * Initialize the navigation stack to have zero entries.
+   */
+  constructor() {
+    this.navigateBackStack = [];
+    this.currPalette = null;
+  }
+
+  /**
+   * Report if the navigation stack is empty.
+   * @return: `true` if the stack is empty; `false` otherwise.
+   */
+  isEmpty () {
+    return this.navigateBackStack.length === 0;
+  }
+
+  /**
+   * Puah a palette onto the top of the navigation stack.
+   * @param: {JsonPaletteType} palette - The palette to push.  If `null` or
+   *                                     `undefined`, the navigation stack is
+   *                                      left untouched.
+   */
+  push (palette: JsonPaletteType) {
+    if (!palette) {
+      return;
+    }
+    this.navigateBackStack.push(palette);
+  }
+
+  /**
+   * Pop and return the most recently pushed palette from the top of the
+   * navigation stack.
+   * @return {JsonPaletteType} - reference to the popped palette; null if the stack is
+   *                     empty.
+   */
+  pop () {
+    if (this.isEmpty()) {
+      return null;
+    } else {
+      return this.navigateBackStack.pop();
+    }
+  }
+
+  /**
+   * Return the palette at the top of the stack without changing the stack
+   * tself.  If an index is given, the palette at that index is returned.  Note
```

Comment:
typo: "tself -> itself".

Another question is, if an index of zero denotes the top of the stack, what if someone wants to peek the palette at the bottom of the stack that actually uses the index zero?

Reply:
I added a `peekLast()` function to look at the bottom of the stack.  I first thought that a special index value such as `-1` defined as the constant `LAST_INDEX` could be used, e.g., `const palette = navStack.peek(LAST_INDEX)`, but that felt less transparent than a function for peeking at the bottom of the stack.  So I switched to `peekLast()`.